scapy/contrib/isotp/isotp_soft_socket.py

```diff
@@ -0,0 +1,1065 @@
+        if to_wait > 0:
+            log_runtime.debug("TimeoutScheduler Thread going to sleep @ %f " +
+                              "for %fs", now, to_wait)
+            interrupted = TimeoutScheduler._event.wait(to_wait)
+            new = TimeoutScheduler._time()
+            log_runtime.debug("TimeoutScheduler Thread awake @ %f, slept for" +
+                              " %f, interrupted=%d", new, new - now,
+                              interrupted)
+
+        # Clear the event so that we can wait on it again,
+        # Must be done before doing the callbacks to avoid losing a set().
+        TimeoutScheduler._event.clear()
+
+    @staticmethod
+    def _task():
+        # type: () -> None
+        """Executed in a background thread, this thread will automatically
+        start when the first timeout is added and stop when the last timeout
+        is removed or executed."""
+
+        log_runtime.debug("TimeoutScheduler Thread spawning @ %f",
+                          TimeoutScheduler._time())
+
+        time_empty = None
+
+        try:
+            while 1:
+                handle = TimeoutScheduler._peek_next()
+                if handle is None:
+                    now = TimeoutScheduler._time()
+                    if time_empty is None:
+                        time_empty = now
+                    # 100 ms of grace time before killing the thread
+                    if TimeoutScheduler.GRACE < now - time_empty:
+                        return
+                TimeoutScheduler._wait(handle)
+                TimeoutScheduler._poll()
+
+        finally:
+            # Worst case scenario: if this thread dies, the next scheduled
+            # timeout will start a new one
+            log_runtime.debug("TimeoutScheduler Thread dying @ %f",
+                              TimeoutScheduler._time())
+            TimeoutScheduler._thread = None
+
+    @staticmethod
+    def _poll():
+        # type: () -> None
+        """Execute all the callbacks that were due until now"""
+
+        handles = TimeoutScheduler._handles
+        handle = None
+        while 1:
+            with TimeoutScheduler._mutex:
+                now = TimeoutScheduler._time()
+                if len(handles) == 0 or handles[0]._when > now:
+                    # There is nothing to execute yet
+                    return
+
+                # Time complexity is O(log n)
+                handle = heapq.heappop(handles)
+                callback = None
+                if handle is not None:
+                    callback = handle._cb
+                    handle._cb = True
+
+            # Call the callback here, outside of the mutex
+            if callable(callback):
+                try:
+                    callback()
+                except Exception:
+                    traceback.print_exc()
+
+    @staticmethod
+    def _time():
+        # type: () -> float
+        if six.PY2:
+            return time.time()
```

Comment:
`time.monitonic()` equivalent on python 2 would be `time.clock()`

Reply:
thanks!